{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8353057a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dash import Dash, html, dcc, dash_table\n",
    "import pandas as pd\n",
    "from dash.dependencies import Input, Output, State, ALL\n",
    "import dash_bootstrap_components as dbc\n",
    "import umap.umap_ as umap\n",
    "from sklearn.cluster import KMeans\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "app = Dash(__name__)\n",
    "#---------------------------------------------------------------------------------------------------\n",
    "path = 'path_to_input_file.csv'\n",
    "df = pd.read_csv(path, na_values='NULL')\n",
    "df.set_index('your_index', inplace=True)\n",
    "df['your_index'] = range(1, len(df) + 1)\n",
    "PAGE_SIZE = 5\n",
    "total_pages = len(df) // PAGE_SIZE\n",
    "if len(df) % PAGE_SIZE != 0:\n",
    "    total_pages += 1\n",
    "\n",
    "# Set the page_current to the last page if there is data\n",
    "page_current = max(total_pages - 1, 0)\n",
    "# Create input fields for adding new rows\n",
    "new_row_inputs = []\n",
    "#---------------------------------------------------------------------------------------------------\n",
    "table_mutations = dash_table.DataTable(\n",
    "    id='datatable-paging',\n",
    "    columns=[],  # Empty columns initially\n",
    "    page_current=page_current,\n",
    "    page_size=PAGE_SIZE,\n",
    "    page_action='custom',\n",
    "    editable=True,  # Allow editing of the table\n",
    "    row_deletable=True,  # Allow deletion of rows\n",
    ")\n",
    "\n",
    "checkbox_categories = {\n",
    "    \"Clinical and laboratory values\": ['AGE', 'SEX', 'FEV', 'EXAML', 'WBC', 'HB', 'PLT', 'LDH', 'BMB', 'PBB'],\n",
    "    \"Cytogenetics\": ['CGNK', 'CGCX', 'inv16t16.16', 't8.21', 't.6.9..p23.q34.', 'inv.3..q21.q26.2.',\n",
    "                  'minus.5', 'del.5q.', 't.9.22..q34.q11.', 'minus.7', 'minus.17', 't.v.11..v.q23.',\n",
    "                  'abn.17p.', 'inv.16.', 't.9.11..p21.23.q23.', 't.8.21.'],\n",
    "    \"Molecular genetics\": ['ASXL1', 'ATRX', 'BCOR', 'BCORL1', 'BRAF', 'CALR', 'CBL', 'CBLB', 'CDKN2A', 'CEBPA',\n",
    "                  'CEBPADM', 'CEBPA.bZIP', 'CEBPA.bZIP.inframe', 'CEBPA.TAD', 'CEBPASTAT', 'CSF3R', 'CUX1',\n",
    "                  'DNMT3A', 'ETV6', 'EZH2', 'FBXW7', 'FLT3I', 'FLT3T', 'FLT3R', 'GATA1', 'GATA2', 'GNAS',\n",
    "                  'HRAS', 'IDH1', 'IDH2', 'IKZF1', 'JAK2', 'KDM6A', 'KIT', 'KRAS', 'MPL', 'MYD88',\n",
    "                  'NOTCH1', 'NPM1', 'NRAS', 'PDGFRA', 'PTEN', 'PHF6', 'PTPN11', 'RAD21', 'RUNX1', 'SETBP1',\n",
    "                  'SF3B1', 'SMC1A', 'SMC3', 'SRSF2', 'STAG2', 'TET2', 'TP53', 'U2AF1', 'WT1', 'ZRSR2']\n",
    "}\n",
    "\n",
    "# Create a list of all features for the header ordering\n",
    "all_features = [feature for sublist in checkbox_categories.values() for feature in sublist]\n",
    "# Create input fields for adding new rows\n",
    "new_row_inputs = []\n",
    "\n",
    "checklist_mutations = html.Div([\n",
    "    html.H5(\"Clinical and laboratory values\"),\n",
    "    dcc.Checklist(\n",
    "        id=\"select_feature_group_1\",\n",
    "        options=[{'label': html.Label('AGE', title='Age of the patients in years'), 'value': 'AGE', 'disabled': False},\n",
    "                 {'label': html.Label('SEX', title='Sex of the patient: 0 for male, 1 for female'), 'value': 'SEX', 'disabled': False},\n",
    "                 {'label': html.Label('FEV', title='FEV transcription factor, ETS family member'), 'value': 'FEV', 'disabled': False},\n",
    "                 {'label': html.Label('EXAML', title=''), 'value': 'EXAML', 'disabled': False},\n",
    "                 {'label': html.Label('WBC', title='White blood cells'), 'value': 'WBC', 'disabled': False},\n",
    "                 {'label': html.Label('HB', title='Hemoglobin'), 'value': 'HB', 'disabled': False},\n",
    "                 {'label': html.Label('PLT', title='Platelet count'), 'value': 'PLT', 'disabled': False},\n",
    "                 {'label': html.Label('LDH', title='Lactate dehydrogenase'), 'value': 'LDH', 'disabled': False},\n",
    "                 {'label': html.Label('BMB', title='Bone marrow biopsies'), 'value': 'BMB', 'disabled': False},\n",
    "                 {'label': html.Label('PBB', title='Polybrominated Biphenyls'), 'value': 'PBB', 'disabled': False}\n",
    "                 ],\n",
    "        value=['AGE', 'PBB', 'BMB', 'PLT'],\n",
    "        labelStyle=dict(display='inline')\n",
    "    ),\n",
    "    html.H5(\"Cytogenetics\"),\n",
    "    dcc.Checklist(\n",
    "        id=\"select_feature_group_2\",\n",
    "        options=[{'label': x, 'value': x, 'disabled': False} for x in checkbox_categories[\"Cytogenetics\"]],\n",
    "        value=[],\n",
    "        labelStyle=dict(display='inline')\n",
    "    ),\n",
    "    html.H5(\"Molecular genetics\"),\n",
    "    dcc.Checklist(\n",
    "        id=\"select_feature_group_3\",\n",
    "        options=[{'label': html.Label('ASXL1', title='ASXL transcriptional regulator 1'), 'value': 'ASXL1', 'disabled': False},\n",
    "                 {'label': html.Label('ATRX', title='ATRX chromatin remodeler'), 'value': 'ATRX', 'disabled': False},\n",
    "                 {'label': html.Label('BCOR', title='BCL6 corepressor'), 'value': 'BCOR', 'disabled': False},\n",
    "                 {'label': html.Label('BCORL1', title='BCL6 corepressor like 1'), 'value': 'BCORL1', 'disabled': False},\n",
    "                 {'label': html.Label('BRAF', title='B-Raf proto-oncogene, serine/threonine kinase'), 'value': 'BRAF', 'disabled': False},\n",
    "                 {'label': html.Label('CALR', title='calreticulin'), 'value': 'CALR', 'disabled': False},\n",
    "                 {'label': html.Label('CBL', title='Cbl proto-oncogene'), 'value': 'CBL', 'disabled': False},\n",
    "                 {'label': html.Label('CBLB', title='Cbl proto-oncogene B'), 'value': 'CBLB', 'disabled': False},\n",
    "                 {'label': html.Label('CDKN2A', title='cyclin dependent kinase inhibitor 2A'), 'value': 'CDKN2A', 'disabled': False},\n",
    "                 {'label': html.Label('CEBPA', title='CCAAT enhancer binding protein alpha'), 'value': 'CEBPA', 'disabled': False},\n",
    "                 {'label': html.Label('CEBPADM', title=''), 'value': 'CEBPADM', 'disabled': False},\n",
    "                 {'label': html.Label('CEBPA.bZIP', title=''), 'value': 'CEBPA.bZIP', 'disabled': False},\n",
    "                 {'label': html.Label('CEBPA.bZIP.inframe', title=''), 'value': 'CEBPA.bZIP.inframe', 'disabled': False},\n",
    "                 {'label': html.Label('CEBPA.TAD', title=''), 'value': 'CEBPA.TAD', 'disabled': False},\n",
    "                 {'label': html.Label('CEBPASTAT', title=''), 'value': 'CEBPASTAT', 'disabled': False},\n",
    "                 {'label': html.Label('CSF3R', title='colony stimulating factor 3 receptor'), 'value': 'CSF3R', 'disabled': False},\n",
    "                 {'label': html.Label('CUX1', title='cut like homeobox 1'), 'value': 'CUX1', 'disabled': False},\n",
    "                 {'label': html.Label('DNMT3A', title='DNA methyltransferase 3 alpha'), 'value': 'DNMT3A', 'disabled': False},\n",
    "                 {'label': html.Label('ETV6', title='ETS variant transcription factor 6'), 'value': 'ETV6', 'disabled': False},\n",
    "                 {'label': html.Label('EZH2', title='enhancer of zeste 2 polycomb repressive complex 2 subunit'), 'value': 'EZH2', 'disabled': False},\n",
    "                 {'label': html.Label('FBXW7', title='F-box and WD repeat domain containing 7'), 'value': 'FBXW7', 'disabled': False},\n",
    "                 {'label': html.Label('FLT3I', title=''), 'value': 'FLT3I', 'disabled': False},\n",
    "                 {'label': html.Label('FLT3T', title=''), 'value': 'FLT3T', 'disabled': False},\n",
    "                 {'label': html.Label('FLT3R', title=''), 'value': 'FLT3R', 'disabled': False},\n",
    "                 {'label': html.Label('GATA1', title='GATA binding protein 1'), 'value': 'GATA1', 'disabled': False},\n",
    "                 {'label': html.Label('GATA2', title='GATA binding protein 2'), 'value': 'GATA2', 'disabled': False},\n",
    "                 {'label': html.Label('GNAS', title='GNAS complex locus'), 'value': 'GNAS', 'disabled': False},\n",
    "                 {'label': html.Label('HRAS', title='HRas proto-oncogene, GTPase'), 'value': 'HRAS', 'disabled': False},\n",
    "                 {'label': html.Label('IDH1', title='isocitrate dehydrogenase (NADP(+)) 1'), 'value': 'IDH1', 'disabled': False},\n",
    "                 {'label': html.Label('IDH2', title='isocitrate dehydrogenase (NADP(+)) 2'), 'value': 'IDH2', 'disabled': False},\n",
    "                 {'label': html.Label('IKZF1', title='IKAROS family zinc finger 1'), 'value': 'IKZF1', 'disabled': False},\n",
    "                 {'label': html.Label('JAK2', title='Janus kinase 2'), 'value': 'JAK2', 'disabled': False},\n",
    "                 {'label': html.Label('KDM6A', title='lysine demethylase 6A'), 'value': 'KDM6A', 'disabled': False},\n",
    "                 {'label': html.Label('KIT', title='KIT proto-oncogene, receptor tyrosine kinase'), 'value': 'KIT', 'disabled': False},\n",
    "                 {'label': html.Label('KRAS', title='KRAS proto-oncogene, GTPase'), 'value': 'KRAS', 'disabled': False},\n",
    "                 {'label': html.Label('MPL', title='MPL proto-oncogene, thrombopoietin receptor'), 'value': 'MPL', 'disabled': False},\n",
    "                 {'label': html.Label('MYD88', title='MYD88 innate immune signal transduction adaptor'), 'value': 'MYD88', 'disabled': False},\n",
    "                 {'label': html.Label('NOTCH1', title='notch receptor 1'), 'value': 'NOTCH1', 'disabled': False},\n",
    "                 {'label': html.Label('NPM1', title='nucleophosmin 1'), 'value': 'NPM1', 'disabled': False},\n",
    "                 {'label': html.Label('NRAS', title='NRAS proto-oncogene, GTPase'), 'value': 'NRAS', 'disabled': False},\n",
    "                 {'label': html.Label('PDGFRA', title='platelet derived growth factor receptor alpha'), 'value': 'PDGFRA', 'disabled': False},\n",
    "                 {'label': html.Label('PTEN', title='phosphatase and tensin homolog'), 'value': 'PTEN', 'disabled': False},\n",
    "                 {'label': html.Label('PHF6', title='PHD finger protein 6'), 'value': 'PHF6','disabled': False},\n",
    "                 {'label': html.Label('PTPN11', title='protein tyrosine phosphatase non-receptor type 11'), 'value': 'PTPN11', 'disabled': False},\n",
    "                 {'label': html.Label('RAD21', title='RAD21 cohesin complex component'), 'value': 'RAD21', 'disabled': False},\n",
    "                 {'label': html.Label('RUNX1', title='RUNX family transcription factor 1'), 'value': 'RUNX1', 'disabled': False},\n",
    "                 {'label': html.Label('SETBP1', title='SET binding protein 1'), 'value': 'SETBP1', 'disabled': False},\n",
    "                 {'label': html.Label('SF3B1', title='splicing factor 3b subunit 1'), 'value': 'SF3B1', 'disabled': False},\n",
    "                 {'label': html.Label('SMC1A', title='structural maintenance of chromosomes 1A'), 'value': 'SMC1A', 'disabled': False},\n",
    "                 {'label': html.Label('SMC3', title='structural maintenance of chromosomes 3'), 'value': 'SMC3', 'disabled': False},\n",
    "                 {'label': html.Label('SRSF2', title='serine and arginine rich splicing factor 2'), 'value': 'SRSF2', 'disabled': False},\n",
    "                 {'label': html.Label('STAG2', title='stromal antigen 2'), 'value': 'STAG2', 'disabled': False},\n",
    "                 {'label': html.Label('TET2', title='tet methylcytosine dioxygenase 2'), 'value': 'TET2', 'disabled': False},\n",
    "                 {'label': html.Label('TP53', title='tumor protein p53'), 'value': 'TP53', 'disabled': False},\n",
    "                 {'label': html.Label('U2AF1', title='U2 small nuclear RNA auxiliary factor 1'), 'value': 'U2AF1', 'disabled': False},\n",
    "                 {'label': html.Label('WT1', title='WT1 transcription factor'), 'value': 'WT1', 'disabled': False},\n",
    "                 {'label': html.Label('ZRSR2', title='zinc finger CCCH-type, RNA binding motif and serine/arginine rich 2'), 'value': 'ZRSR2', 'disabled': False}\n",
    "                 ],\n",
    "        value=['DNMT3A', 'KIT', 'CGCX', 'NPM1'],\n",
    "        labelStyle=dict(display='inline')\n",
    "    )\n",
    "])\n",
    "\n",
    "UMAP = dcc.Graph(id='Umap_projection2d_live_adding', figure={})\n",
    "dummy_div = html.Div(id='dummy-div', style={'display': 'none'})\n",
    "\n",
    "app.layout = html.Div([\n",
    "    html.H1(\"UMAP Projection for original data\", title=\"With this application you can check to which risk group belong the specific patient.\\n\"\n",
    "                                                       \"You can mark any number of the variables and then write down your results.\\n\"\n",
    "                                                       \"Once you filled the cells with the information just press the button and wait for the results to appear.\\n\"\n",
    "                                                       \"If you want to update the results in the table just go to other page and come back.\\n\"\n",
    "                                                       \"If you do not know what the specific abbrevation mean, just hover on it\",\n",
    "            style={'text-align': 'left'}),\n",
    "    checklist_mutations,\n",
    "    table_mutations,\n",
    "    html.H3(\"Enter new row data:\"),\n",
    "    html.Div(id='new-row-inputs-container'),  # Container for new row input fields\n",
    "    dbc.Button('Add New Row', id='add-new-row-button', n_clicks=0, color='primary'),\n",
    "    dcc.Graph(id='Umap_projection2d_live_adding', figure={}),  # Add the UMAP graph here\n",
    "    dummy_div  # Include the dummy div in the layout\n",
    "])\n",
    "\n",
    "# Callback to dynamically update the new row input fields based on checklist selection\n",
    "@app.callback(\n",
    "    Output('new-row-inputs-container', 'children'),\n",
    "    Input('select_feature_group_1', 'value'),\n",
    "    Input('select_feature_group_2', 'value'),\n",
    "    Input('select_feature_group_3', 'value'))\n",
    "\n",
    "def update_new_row_inputs(group_1_selected, group_2_selected, group_3_selected):\n",
    "    # Concatenate the selected features from all groups\n",
    "    selected_features = group_1_selected + group_2_selected + group_3_selected\n",
    "    # Create input fields for each selected feature with smaller cells\n",
    "    new_row_inputs = [\n",
    "        dbc.Input(\n",
    "            id={'type': 'new-row-input', 'index': feature},\n",
    "            type='text',\n",
    "            placeholder=f'Enter {feature}',\n",
    "            style={'width': '80px'}  # Adjust the width as desired\n",
    "        )\n",
    "        for feature in selected_features\n",
    "    ]\n",
    "    return new_row_inputs\n",
    "\n",
    "# Callback to add a new row to the DataFrame when the \"Add New Row\" button is clicked\n",
    "@app.callback(\n",
    "    Output('dummy-div', 'children'),\n",
    "    Input('add-new-row-button', 'n_clicks'),\n",
    "    State({'type': 'new-row-input', 'index': ALL}, 'value'),\n",
    "    State('select_feature_group_1', 'value'),\n",
    "    State('select_feature_group_2', 'value'),\n",
    "    State('select_feature_group_3', 'value'),\n",
    ")\n",
    "def add_new_row(n_clicks, new_row_values, group_1_selected, group_2_selected, group_3_selected):\n",
    "    if n_clicks > 0:\n",
    "        selected_features = group_1_selected + group_2_selected + group_3_selected\n",
    "        num_features_selected = len(selected_features)\n",
    "        num_input_values = len(new_row_values)\n",
    "        if num_input_values == num_features_selected:\n",
    "            # Create a dictionary for the new row data\n",
    "            new_row = {col: val for col, val in zip(selected_features, new_row_values)}\n",
    "            # Convert the new row to a DataFrame\n",
    "            new_row_df = pd.DataFrame([new_row])\n",
    "            # Concatenate the new row DataFrame to the existing DataFrame\n",
    "            global df\n",
    "            df = pd.concat([df, new_row_df], ignore_index=True)\n",
    "\n",
    "            # Sort DataFrame in reverse order based on the index column (Pat) and reset the index\n",
    "            df = df.sort_values(by='Pat', ascending=False).reset_index(drop=True)\n",
    "\n",
    "    return None\n",
    "\n",
    "@app.callback(\n",
    "    Output('datatable-paging', 'data'),\n",
    "    Output('datatable-paging', 'columns'),\n",
    "    Input('datatable-paging', \"page_current\"),\n",
    "    Input('datatable-paging', \"page_size\"),\n",
    "    Input('select_feature_group_1', 'value'),\n",
    "    Input('select_feature_group_2', 'value'),\n",
    "    Input('select_feature_group_3', 'value'),\n",
    ")\n",
    "def update_table(page_current, page_size, group_1_selected, group_2_selected, group_3_selected):\n",
    "    # Concatenate the selected features from all groups\n",
    "    selected_features = group_1_selected + group_2_selected + group_3_selected\n",
    "    # Filter the DataFrame based on selected features\n",
    "    filtered_df = df[selected_features]\n",
    "    # Generate columns dynamically based on selected features\n",
    "    columns = [{\"name\": col, \"id\": col} for col in filtered_df.columns]\n",
    "\n",
    "    start_index = page_current * page_size\n",
    "    end_index = (page_current + 1) * page_size\n",
    "\n",
    "    # Get the sliced data for the current page\n",
    "    sliced_data = filtered_df.iloc[start_index:end_index].to_dict('records')\n",
    "\n",
    "    return sliced_data, columns\n",
    "\n",
    "@app.callback(\n",
    "    Output('Umap_projection2d_live_adding', 'figure'),\n",
    "    Input('select_feature_group_1', 'value'),\n",
    "    Input('select_feature_group_2', 'value'),\n",
    "    Input('select_feature_group_3', 'value'),\n",
    "    Input('add-new-row-button', 'n_clicks'),  # Add the button as an input\n",
    ")\n",
    "def update_graph(group_1_selected, group_2_selected, group_3_selected, n_clicks):\n",
    "    # Concatenate the selected features from all groups\n",
    "    selected_features = group_1_selected + group_2_selected + group_3_selected\n",
    "    # Filter the DataFrame based on selected features\n",
    "    dff = df[selected_features]\n",
    "    kmeans = KMeans(4)\n",
    "    kmeans.fit(dff)\n",
    "    clusters = dff.copy()\n",
    "    clusters['clusters_pred'] = kmeans.fit_predict(dff)\n",
    "    proj_2d = umap.UMAP(random_state=42).fit_transform(dff)\n",
    "\n",
    "    fig = go.Figure()\n",
    "\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=proj_2d[:, 0],\n",
    "        y=proj_2d[:, 1],\n",
    "        mode='markers',\n",
    "        marker=dict(\n",
    "            size=3,\n",
    "            color=clusters['clusters_pred'],\n",
    "            colorscale='Viridis',\n",
    "            showscale=True\n",
    "        ),\n",
    "        showlegend=False\n",
    "    ))\n",
    "\n",
    "    # Adding a red marker at the last point\n",
    "    fig.add_trace(go.Scatter(\n",
    "        x=[proj_2d[-1, 0]],\n",
    "        y=[proj_2d[-1, 1]],\n",
    "        mode='markers',\n",
    "        marker=dict(size=10, color='red'),\n",
    "        showlegend=False\n",
    "    ))\n",
    "    return fig\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    app.run_server(debug=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
